README.md
## rmote

![gif](https://cloud.githubusercontent.com/assets/1275592/9618810/99cf5b2c-50be-11e5-885d-a4de16919271.gif)

### Running R on a remote server

R users often find themselves needing to log in to a remote machine to do analysis.  Sometimes this is due to data restrictions, computing power on the remote machine, etc.  Users can ssh in and run R in a terminal, but it is not possible to look at graphics, etc.

There are currently three approaches that I am aware of to deal with this:

1. Install [RStudio Server](https://www.rstudio.com/products/rstudio-server-pro/) on the remote server and use that from a web browser on your local machine.  Graphics output is shown in the IDE.
2. Use X11 forwarding (`ssh -X|Y`).  Graphics output is sent back to your machine.
3. Use VNC with a linux desktop environment like KDE or GNOME.

Whenever possible, #1 is by far the best way to go and is one of the beautiful things about RStudio Server.  #2 is not a good choice - plots are very slow to render, the quality is terrible, and it doesn't support recent advances in plot outputs like [htmlwidgets](http://htmlwidgets.org) (unless you launch firefox through X11, which will mean you might get to look at one plot per hour).  #3 is okay if it is available and you are comfortable working in one of these desktop environments.

There could be other obvious ways to deal with this that I am oblivious to.

### A problem

Often we do not have the choice of installing RStudio Server or a desktop environment on the remote machine.  Also, some users prefer to work in a terminal sending commands from a favorite text editor on a local machine, but still want to see graphics.  We would like to have something better than X11 forwarding to view graphics and other output when running R in a terminal on a remote machine.

### A solution

The rmote package is an attempt to make working in R over ssh on a server a bit more pleasant in terms of viewing output.  It uses [servr](https://github.com/yihui/servr) on the remote machine to serve R graphics as they are created.  These can be viewed on the local machine in a web browser. The user's local browser will automatically refresh each time a new output is available.

Currently there is support for lattice, ggplot2, htmlwidgets, and help output.

### Usage

1. Choose a port to run your remote server on (default is 4321)
2. ssh into the remote machine, mapping the port on the remote back to your local machine:

    ```
    ssh -L 4321:localhost:4321 -L 8100:localhost:8100 user@remote
    ```

    I also add port 8100 so I can forward shiny apps back to my local machine on a dedicated port.

3. On the remote machine launch R and install the latest version of servr and rmote (one time only)

    ```r
    install.packages("rmote", repos = c(CRAN = "http://cran.rstudio.com",
      tessera = "http://packages.tessera.io"))
    ```

    or alternatively

    ```r
    devtools::install_github(c("yihui/servr", "hafen/rmote"))
    ```

    Note that this package will probably never be on CRAN since it overwrites some standard R S3 methods.

4. Run the following in R on the remote:

    ```r
    rmote::start_rmote()
    ```

    To view some of the options for this, see `?start_rmote`.  One option is the port, which needs to match the one your forwarded in step 2 (4321 is the default.)

5. On your local machine, open up your web browser to `localhost:4321`

    Now as you create compatible plots on your remote machine, your browser on your local fmachine will automatically update to show the results.  For example, try running each of the following in succession on the server:

    ```r
    ?plot

    qplot(mpg, wt, data=mtcars, colour=cyl)

    dotplot(variety ~ yield | year * site, data=barley)

    library(rbokeh)
    figure() %>% ly_hexbin(rnorm(10000), rnorm(10000))
    ```

This process is slightly more tedious than `ssh -X` for initial setup, but much faster and more functional.

### Other useful utilities

If you must work on a remote terminal, here are some additional utilities that help make things nice:

- [colorout](https://github.com/jalvesaq/colorout)
- [Sublime](https://www.sublimetext.com) + [SFTP](http://wbond.net/sublime_packages/sftp)
- [Vim](http://www.vim.org) + [Vim-R-plugin](https://github.com/vim-scripts/Vim-R-plugin)



NEWS.md
Version 0.3
----------------------------------------------------------------------

- Add history sidebar to viewer and option to use it or not (0.3.4)
- Add history thumbnail functions for all output modes (0.3.4)
- Clean up html and css (0.3.4)
- Add `hostname` option that will show hostname in page title (0.3.3)
- Add preliminary tests (0.3.3)
- Ensure raster graphics don't render if someone/thing has explicitly opened their own device (0.3.3)
- Fix iframe refresh issue (thanks @yihui) (0.3.2)
- Add base R graphics support (0.3.1)
- Add `rmote_on()` and `rmote_off()`
- Add main index page with history

Version 0.2
----------------------------------------------------------------------

- Add `capabilities()` checking for png
- When the rmote server isn't running, restore default behavior of print methods
- Add `rmote_device()` function to allow control over plot dimensions, etc.
- Add `daemon = FALSE` option for potentially running a single dedicated rmote server for multiple R sessions on a remote server
- Add `rmote_mode()` to toggle rmote on / off (useful for `daemon = FALSE` option)
- Change start and stop to `rmote_start()` and `rmote_stop()`
- Make graphics look good on retina displays by default
- Use servr's new `httw()` function for live reloading
- Remove packages from depends and import
- Avoid use of `:::`


tests/testthat.R
library(testthat)

test_check("rmote")


tests/testthat/test-package.R
rmdr <- file.path(tempdir(), "rmote_server")
f1 <- tempfile(fileext = ".png")
f2 <- tempfile(fileext = ".png")
f3 <- tempfile(fileext = ".png")

rmote::start_rmote(rmdr)

?lm
r1 <- file.exists(file.path(rmdr, "output0001.html"))

?glm
r2 <- file.exists(file.path(rmdr, "output0002.html"))

library(lattice)
xyplot(1:10 ~ 1:10)
r3 <- file.exists(file.path(rmdr, "output0003.html"))

png(file = f1)
xyplot(1:10 ~ 1:10)
dev.off()
r4 <- file.exists(f1)

library(ggplot2)
qplot(mpg, wt, data = mtcars)
r5 <- file.exists(file.path(rmdr, "output0004.html"))

png(file = f2)
qplot(mpg, wt, data = mtcars)
dev.off()
r6 <- file.exists(f1)

plot(1:10)
rmote::plot_done()
r7 <- file.exists(file.path(rmdr, "output0005.html"))

png(file = f3)
plot(1:10)
dev.off()
r8 <- file.exists(f3)

# TODO: testthat doesn't like this
# runs fine in .GlobalEnv in an interactive session
# look into it...
test_that("rmote working", {
  if(FALSE) {
    expect_true(r1)
    expect_true(r2)
    expect_true(r3)
    expect_true(r4)
    expect_true(r5)
    expect_true(r6)
    expect_true(r7)
    expect_true(r8)
  }
})

rmote::stop_rmote()


R/help.R

#' Print help page to servr
#'
#' @param x help object
#' @param \ldots  additional parameters
#' @S3method print help_files_with_topic
#' @importFrom tools Rd2HTML
print.help_files_with_topic <- function(x, ...) {

  file <- as.character(x)

  help_opt <- getOption("rmote_help", FALSE)

  if(is_rmote_on() && help_opt && length(file) == 1 && grepl("/help/", file)) {
    message("serving help through rmote")

    res <- try({
      topic <- gsub(".*/help/(.*)", "\\1", file)
      package <- gsub(".*/(.*)/help/.*", "\\1", file)
      rdb_path <- file.path(system.file("help", package = package), package)
      tmp <- getFromNamespace("fetchRdDB", "tools")(rdb_path, topic)
      capture.output(tools::Rd2HTML(tmp))
    }, silent = TRUE)
    if(!inherits(res, "try-error")) {
      server_dir <- get_server_dir()
      # move R.css over
      # file.path(R.home('doc'), 'html', 'R.css')
      if(!file.exists(file.path(server_dir, "R.css")))
        file.copy(file.path(system.file(package = "rmote"), "R.css"), server_dir)

      ii <- get_output_index()
      writeLines(c("<!-- DISABLE-SERVR-WEBSOCKET -->", res),
        file.path(server_dir, get_output_file(ii)))
      write_index(ii)

      if(is_history_on()) {
        message("making thumbnail")
        fbase <- file.path(server_dir, "thumbs")
        if(!file.exists(fbase))
          dir.create(fbase)
        nf <- file.path(fbase, gsub("html$", "png", get_output_file(ii)))
        opts <- list(filename = nf, width = 300, height = 150)
        if(capabilities("cairo"))
          opts$type <- "cairo-png"
        do.call(png, opts)
        getFromNamespace("print.trellis", "lattice")(text_plot(paste("help:", topic)))
        dev.off()
      }

      return()
    }
  } else {
    getFromNamespace("print.help_files_with_topic", "utils")(x)
  }
}


R/misc.R

get_hostname <- function() {
  hn <- try(system("hostname", intern = TRUE, ignore.stderr = TRUE), silent = TRUE)
  if(inherits(hn, "try-error"))
    hn <- ""
  hn
}

get_output_index <- function() {
  server_dir <- get_server_dir()
  idx_path <- file.path(get_server_dir(), ".idx")
  if(!file.exists(idx_path))
    return(1)
  as.integer(readLines(idx_path, warn = FALSE)[1]) + 1
}

get_output_file <- function(ii) {
  sprintf("output%04d.html", ii)
}

set_index_template <- function() {
  a <- readLines(system.file("index.html", package = "rmote"), warn = FALSE)
  if(getOption("rmote_hostname")) {
    idx <- grepl("rmote viewer", a[1:10])
    hn <- get_hostname()
    if(length(idx) > 0 && hn != "")
      a[idx] <- gsub("rmote viewer", paste("rmote:", hn), a[idx])
  }
  options(rmote_index = paste(c(a, ""), collapse = "\n"))
}

write_html <- function(html) {
  ii <- get_output_index()
  fp <- file.path(get_server_dir(), get_output_file(ii))
  htmltools::save_html(
    tagList(HTML("<!-- DISABLE-SERVR-WEBSOCKET -->"), html),
    file = fp)
  write_index(ii)
  fp
}

write_index <- function(ii) {
  if(is.null(getOption("rmote_index")))
    set_index_template()
  res <- gsub("___max_page___", ii, getOption("rmote_index"))
  res <- gsub("___history___", ifelse(is_history_on(), "true", "false"), res)
  # write out file index
  cat(ii, file = file.path(get_server_dir(), ".idx"))
  cat(res, file = file.path(get_server_dir(), "index.html"))
}

is_rmote_on <- function() {
  getOption("rmote_on", FALSE) || length(servr::daemon_list()) > 0
}

is_history_on <- function() {
  getOption("rmote_history", TRUE)
}

no_other_devices <- function() {
  res <- length(dev.list()) == 0
  if(!res) {
    message("- not sending to rmote because another graphics device has been opened...")
    message("- sending to the open graphics device instead...")
    message("- to send to rmote, close all active graphics devices using graphics.off()")
  }
  res
}

get_server_dir <- function() {
  server_dir <- getOption("rmote_server_dir")
  if(is.null(server_dir))
    stop("No setting for rmote_server_dir - make sure to call rmote_server_init()")

  if(!file.exists(server_dir))
    dir.create(server_dir, recursive = TRUE, showWarnings = FALSE)

  server_dir
}

get_port <- function() {
  port <- getOption("rmote_server_port")
  if(is.null(port))
    stop("No setting for rmote_server_port - make sure to call rmote_server_init()")

  port
}

text_plot <- function(text) {
  xyplot(NA ~ NA, xlab = "", ylab = "",
    scales = list(draw = FALSE),
    panel = function(x, y, ...)
    panel.text(0.5, 0.5, text, cex = 2.5))
}

#' @importFrom png readPNG
#' @importFrom graphics rasterImage
make_thumb <- function(in_file, out_file, width, height) {
  fbase <- dirname(out_file)
  if(!dir.exists(fbase))
    dir.create(fbase)

  max_height <- 150
  ratio <- max_height / height
  height <- ratio * height
  width <- ratio * width

  img <- png::readPNG(in_file)
  opts <- list(filename = out_file, width = width, height = height)
  if(capabilities("cairo"))
    opts$type <- "cairo-png"
  do.call(png, opts)
    par(mar = c(0,0,0,0), xaxs = "i", yaxs = "i", ann = FALSE)
    plot(1:2, type = "n", xaxt = "n", yaxt = "n", xlab = "", ylab = "")
    lim <- par()
    graphics::rasterImage(img, lim$usr[1], lim$usr[3], lim$usr[2], lim$usr[4])
  dev.off()
}

dir.exists <- function(x) {
  if(file.exists(x) & file.info(x)$isdir)
    return(TRUE)
  return(FALSE)
}


R/rmote.R
#' @import ggplot2
#' @import lattice
#' @import htmlwidgets
NULL

#' Initialize a remote servr
#'
#' @param server_dir directory to launch servr from
#' @param port port to run servr on
#' @param daemon logical - should the server be started as a daemon?
#' @param help (logical) send results of `?` to servr
#' @param graphics (logical) send traditional lattice / ggplot2 plots to servr
#' @param basegraphics (logical) send base graphics to servr
#' @param htmlwidgets (logical) send htmlwidgets to servr
#' @param hostname (logical) try to get hostname and use it in viewer page title
#' @param history (logical) should history thumbnails be created and shown in the viewer?
#'
#' @export
#' @importFrom servr httd
start_rmote <- function(
  server_dir = file.path(tempdir(), "rmote_server"),
  port = 4321, daemon = TRUE,
  help = TRUE, graphics = TRUE,
  basegraphics = TRUE, htmlwidgets = TRUE,
  hostname = TRUE, history = TRUE) {

  if(!file.exists(server_dir))
    dir.create(server_dir, recursive = TRUE, showWarnings = FALSE)

  options(rmote_server_dir = server_dir)
  options(rmote_server_port = port)
  options(rmote_help = help)
  options(rmote_graphics = graphics)
  options(rmote_htmlwidgets = htmlwidgets)
  options(rmote_basegraphics = basegraphics)
  options(rmote_hostname = hostname)
  options(rmote_history = history)

  set_index_template()

  if(basegraphics)
    set_base_plot_hook()

  try(servr::httw(server_dir, pattern = "index.html", port = port,
      daemon = daemon, browser = FALSE), silent = TRUE)
}

#' Stop an rmote server
#' @export
stop_rmote <- function() {
  plot_done()
  if(getOption("rmote_basegraphics", FALSE))
    unset_base_plot_hook()
  options(rmote_on = FALSE)
  servr::daemon_stop()
}

#' Set the rmote mode to be on
#'
#' @param server_dir directory where rmote server is running
#' @param help (logical) send results of `?` to servr
#' @param graphics (logical) send traditional lattice / ggplot2 plots to servr
#' @param basegraphics (logical) send base graphics to servr
#' @param htmlwidgets (logical) send htmlwidgets to servr
#' @param hostname (logical) try to get hostname and use it in viewer page title
#' @param history (logical) should history thumbnails be created and shown in the viewer?
#' @note This is useful when running multiple R sessions on a server, where all will serve the same rmote process.  It is not necessary to call this in the same session on which \code{\link{start_rmote}} has been called, but on any other R sessions.
#' @export
rmote_on <- function(server_dir,
  help = TRUE, graphics = TRUE,
  basegraphics = TRUE, htmlwidgets = TRUE,
  hostname = TRUE, history = TRUE
) {

  if(!file.exists(server_dir))
    stop("The location of server_dir does not exist - no rmote server running here...")

  options(rmote_on = TRUE)

  options(rmote_server_dir = server_dir)
  options(rmote_help = help)
  options(rmote_graphics = graphics)
  options(rmote_htmlwidgets = htmlwidgets)
  options(rmote_basegraphics = basegraphics)
  options(rmote_hostname = hostname)
  options(rmote_history = history)

  if(basegraphics)
    set_base_plot_hook()

  invisible(NULL)
}

#' Set the rmote mode to be off
#'
#' @export
rmote_off <- function() {
  plot_done()
  if(getOption("rmote_basegraphics", FALSE))
    unset_base_plot_hook()
  options(rmote_on = FALSE)
}


R/trellis.R
#' Print trellis plot to servr
#'
#' @param x trellis object
#' @param \ldots  additional parameters
#' @S3method print trellis
#' @import htmltools
#' @importFrom digest digest
print.trellis <- function(x, ...) {
  print_graphics(x)
}

#' Print ggplot2 plot to servr
#'
#' @param x ggplot object
#' @param \ldots  additional parameters
#' @S3method print ggplot
print.ggplot <- function(x, ...) {
  print_graphics(x)
}

print_graphics <- function(x) {

  graphics_opt <- getOption("rmote_graphics", FALSE)

  if(is_rmote_on() && graphics_opt && no_other_devices()) {
    message("serving graphics through rmote")

    output_dir <- file.path(get_server_dir(), "plots")
    if(!file.exists(output_dir))
      dir.create(output_dir, recursive = TRUE)

    opts <- getOption("rmote_device")
    if(is.null(opts)) {
      rmote_device()
      opts <- getOption("rmote_device")
    }

    if(is.null(opts$filename)) {
      plot_base <- digest::digest(x)
    } else {
      plot_base <- opts$filename
    }
    file <- file.path("plots", paste0(plot_base, ".", opts$type))
    ofile <- file.path(output_dir, paste0(plot_base, ".", opts$type))

    if(opts$type == "png") {
      ww <- opts$width
      hh <- opts$height
      if(opts$retina) {
        opts$width <- opts$width * 2
        opts$height <- opts$height * 2
        opts$res <- 150
      }
      html <- tags$html(
        tags$head(tags$title(paste("raster plot:", plot_base))),
        tags$body(tags$img(src = file,
          width = paste0(ww, "'x"), height = paste0(hh, "'x"))))
      opts$type <- NULL
      opts$retina <- NULL
      opts$filename <- ofile
      if(capabilities("cairo"))
        opts$type <- "cairo-png"
      do.call(png, opts)
    } else if(opts$type == "pdf") {
      html <- tags$html(
        tags$head(tags$title(paste("raster plot:", plot_base))),
        tags$body(tags$a(href = file, "pdf", target = "_blank")))
      opts$type <- NULL
      opts$file <- ofile
      do.call(pdf, opts)
    }

    if(inherits(x, c("trellis", "ggplot"))) {
      if(inherits(x, "trellis"))
        getFromNamespace("print.trellis", "lattice")(x)
      if(inherits(x, "ggplot"))
        getFromNamespace("print.ggplot", "ggplot2")(x)
      dev.off()

      write_html(html)
    } else if(inherits(x, "base_graphics")) {
      message("when finished with plot commands, call plot_done()")
      options(rmote_baseplot = html)
    }

    return()
  } else {
    if(inherits(x, "trellis"))
      return(getFromNamespace("print.trellis", "lattice")(x))
    if(inherits(x, "ggplot"))
      return(getFromNamespace("print.ggplot", "ggplot2")(x))
  }
}

#' Set device parameters for traditional grahpics plot output
#'
#' @param type either "png" or "pdf"
#' @param filename optional name for file (should have no extension and no directories)
#' @param retina if TRUE and type is "png", the png file will be plotted at twice its size and shown at the original size, to look better on high resolution displays
#' @param \ldots parameters passed on to either \code{\link[grDevices]{png}} or \code{\link[grDevices]{pdf}} (such as width, height, etc.)
#' @export
rmote_device <- function(type = c("png", "pdf"), filename = NULL, retina = TRUE, ...) {

  type <- match.arg(type)
  opts <- list(...)
  opts$type <- type
  opts$filename <- filename
  if(type == "png") {
    opts$retina <- retina
    if(is.null(opts$width))
      opts$width <- 480
    if(is.null(opts$height))
      opts$height <- 480
  }

  options(rmote_device = opts)
}







R/htmlwidgets.R

#' Print an htmlwidget to servr
#'
#' @param x htmlwidget object
#' @param \ldots  additional parameters
#' @S3method print htmlwidget
print.htmlwidget <- function(x, ...) {

  widget_opt <- getOption("rmote_htmlwidgets", FALSE)

  if(is_rmote_on() && widget_opt) {
    message("serving htmlwidgets through rmote")

    res <- try({
      html <- htmltools::as.tags(x, standalone = TRUE)
      write_html(html)
    })

    # make thumbnail of htmlwidget
    if(is_history_on()) {
      message("making thumbnail")
      fbase <- file.path(get_server_dir(), "thumbs")
      nf <- file.path(fbase, gsub("html$", "png", basename(res)))
      if(!inherits(res, "try-error")) {
        width <- x$width
        height <- x$height
        x$sizingPolicy$padding <- 0
        if(is.null(width)) width <- 600
        if(is.null(height)) height <- 400

        tf <- tempfile(fileext = ".png")
        ws_res <- try(webshot::webshot(paste0("file://", res), file = tf,
          selector = ".html-widget"), silent = TRUE)
        if(!inherits(ws_res, "try-error")) {
          suppressMessages(make_thumb(tf, nf, width = width, height = height))
        } else {
          opts <- list(filename = nf, width = 300, height = 150)
          if(capabilities("cairo"))
            opts$type <- "cairo-png"
          do.call(png, opts)
          getFromNamespace("print.trellis", "lattice")(text_plot("htmlwidget"))
          dev.off()
        }
      }
      return()
    }

  } else {
    getFromNamespace("print.htmlwidget", "htmlwidgets")(x)
  }
}



R/basegraphics.R

#' Let rmote know that a base R plot is complete and ready to serve
#' @export
plot_done <- function() {
  make_base_plot()
  invisible(NULL)
}

make_base_plot <- function() {
  rmb <- getOption("rmote_baseplot")
  if(!is.null(rmb)) {
    dev.off()
    res <- write_html(rmb$html)
    options(rmote_baseplot = NULL)
    if(is_history_on())
      make_raster_thumb(res, rmb$cur_type, rmb$opts, rmb$ofile)
  }
}

set_base_plot_hook <- function() {
  options(prev_plot_hook = getHook("before.plot.new"))
  setHook("before.plot.new", function() {
    # if a device was opened up automatically, turn it off
    # (automatic devices don't have a path)
    fp <- attr(.Device, "filepath")
    if(is.null(fp) || fp == "Rplots.pdf")
      dev.off()

    # in case previous plot has never finished
    getFromNamespace("make_base_plot", "rmote")()

    # this will call png or pdf with appropriate options
    dummy <- structure(list(Sys.time()), class = "base_graphics")
    getFromNamespace("print_graphics", "rmote")(dummy)
  }, "replace")
}

unset_base_plot_hook <- function() {
  setHook("before.plot.new", getOption("pre_plot_hook"), "replace")
}



inst/live/live.js
 /*
  Live.js - One script closer to Designing in the Browser
  Written for Handcraft.com by Martin Kool (@mrtnkl).

  Version 4.
  Recent change: Made stylesheet and mimetype checks case insensitive.

  http://livejs.com
  http://livejs.com/license (MIT)
  @livejs

  Include live.js#css to monitor css changes only.
  Include live.js#js to monitor js changes only.
  Include live.js#html to monitor html changes only.
  Mix and match to monitor a preferred combination such as live.js#html,css

  By default, just include live.js to monitor all css, js and html changes.

  Live.js can also be loaded as a bookmarklet. It is best to only use it for CSS then,
  as a page reload due to a change in html or css would not re-include the bookmarklet.
  To monitor CSS and be notified that it has loaded, include it as: live.js#css,notify
*/
(function () {

  var headers = { "Etag": 1, "Last-Modified": 1, "Content-Length": 1, "Content-Type": 1 },
      resources = {},
      pendingRequests = {},
      currentLinkElements = {},
      oldLinkElements = {},
      interval = 1000,
      loaded = false,
      active = { "html": 1, "css": 1, "js": 1 };

  var Live = {

    // performs a cycle per interval
    heartbeat: function () {
      if (document.body) {
        // make sure all resources are loaded on first activation
        if (!loaded) Live.loadresources();
        Live.checkForChanges();
      }
      setTimeout(Live.heartbeat, interval);
    },

    // loads all local css and js resources upon first activation
    loadresources: function () {

      // helper method to assert if a given url is local
      function isLocal(url) {
        var loc = document.location,
            reg = new RegExp("^\\.|^\/(?!\/)|^[\\w]((?!://).)*$|" + loc.protocol + "//" + loc.host);
        return url.match(reg);
      }

      // gather all resources
      var scripts = document.getElementsByTagName("script"),
          links = document.getElementsByTagName("link"),
          uris = [];

      // track local js urls
      for (var i = 0; i < scripts.length; i++) {
        var script = scripts[i], src = script.getAttribute("src");
        if (src && isLocal(src))
          uris.push(src);
        if (src && src.match(/\blive.js#/)) {
          for (var type in active)
            active[type] = src.match("[#,|]" + type) != null
          if (src.match("notify"))
            alert("Live.js is loaded.");
        }
      }
      if (!active.js) uris = [];
      if (active.html) uris.push(document.location.href);

      // track local css urls
      for (var i = 0; i < links.length && active.css; i++) {
        var link = links[i], rel = link.getAttribute("rel"), href = link.getAttribute("href", 2);
        if (href && rel && rel.match(new RegExp("stylesheet", "i")) && isLocal(href)) {
          uris.push(href);
          currentLinkElements[href] = link;
        }
      }

      // initialize the resources info
      for (var i = 0; i < uris.length; i++) {
        var url = uris[i];
        Live.getHead(url, function (url, info) {
          resources[url] = info;
        });
      }

      // add rule for morphing between old and new css files
      var head = document.getElementsByTagName("head")[0],
          style = document.createElement("style"),
          rule = "transition: all .3s ease-out;"
      css = [".livejs-loading * { ", rule, " -webkit-", rule, "-moz-", rule, "-o-", rule, "}"].join('');
      style.setAttribute("type", "text/css");
      head.appendChild(style);
      style.styleSheet ? style.styleSheet.cssText = css : style.appendChild(document.createTextNode(css));

      // yep
      loaded = true;
    },

    // check all tracking resources for changes
    checkForChanges: function () {
      for (var url in resources) {
        if (pendingRequests[url])
          continue;

        Live.getHead(url, function (url, newInfo) {
          var oldInfo = resources[url],
              hasChanged = false;
          resources[url] = newInfo;
          for (var header in oldInfo) {
            // do verification based on the header type
            var oldValue = oldInfo[header],
                newValue = newInfo[header],
                contentType = newInfo["Content-Type"];
            switch (header.toLowerCase()) {
              case "etag":
                if (!newValue) break;
                // fall through to default
              default:
                hasChanged = oldValue != newValue;
                break;
            }
            // if changed, act
            if (hasChanged) {
              Live.refreshResource(url, contentType);
              break;
            }
          }
        });
      }
    },

    // act upon a changed url of certain content type
    refreshResource: function (url, type) {
      switch (type.toLowerCase()) {
        // css files can be reloaded dynamically by replacing the link element
        case "text/css":
          var link = currentLinkElements[url],
              html = document.body.parentNode,
              head = link.parentNode,
              next = link.nextSibling,
              newLink = document.createElement("link");

          html.className = html.className.replace(/\s*livejs\-loading/gi, '') + ' livejs-loading';
          newLink.setAttribute("type", "text/css");
          newLink.setAttribute("rel", "stylesheet");
          newLink.setAttribute("href", url + "?now=" + new Date() * 1);
          next ? head.insertBefore(newLink, next) : head.appendChild(newLink);
          currentLinkElements[url] = newLink;
          oldLinkElements[url] = link;

          // schedule removal of the old link
          Live.removeoldLinkElements();
          break;

        // check if an html resource is our current url, then reload
        case "text/html":
          if (url != document.location.href)
            return;

          // local javascript changes cause a reload as well
        case "text/javascript":
        case "application/javascript":
        case "application/x-javascript":
          document.location.reload();
      }
    },

    // removes the old stylesheet rules only once the new one has finished loading
    removeoldLinkElements: function () {
      var pending = 0;
      for (var url in oldLinkElements) {
        // if this sheet has any cssRules, delete the old link
        try {
          var link = currentLinkElements[url],
              oldLink = oldLinkElements[url],
              html = document.body.parentNode,
              sheet = link.sheet || link.styleSheet,
              rules = sheet.rules || sheet.cssRules;
          if (rules.length >= 0) {
            oldLink.parentNode.removeChild(oldLink);
            delete oldLinkElements[url];
            setTimeout(function () {
              html.className = html.className.replace(/\s*livejs\-loading/gi, '');
            }, 100);
          }
        } catch (e) {
          pending++;
        }
        if (pending) setTimeout(Live.removeoldLinkElements, 50);
      }
    },

    // performs a HEAD request and passes the header info to the given callback
    getHead: function (url, callback) {
      pendingRequests[url] = true;
      var xhr = window.XMLHttpRequest ? new XMLHttpRequest() : new ActiveXObject("Microsoft.XmlHttp");
      xhr.open("HEAD", url, true);
      xhr.onreadystatechange = function () {
        delete pendingRequests[url];
        if (xhr.readyState == 4 && xhr.status != 304) {
          xhr.getAllResponseHeaders();
          var info = {};
          for (var h in headers) {
            var value = xhr.getResponseHeader(h);
            // adjust the simple Etag variant to match on its significant part
            if (h.toLowerCase() == "etag" && value) value = value.replace(/^W\//, '');
            if (h.toLowerCase() == "content-type" && value) value = value.replace(/^(.*?);.*?$/i, "$1");
            info[h] = value;
          }
          callback(url, info);
        }
      }
      xhr.send();
    }
  };

  // start listening
  if (document.location.protocol != "file:") {
    if (!window.liveJsLoaded)
      Live.heartbeat();

    window.liveJsLoaded = true;
  }
  else if (window.console)
    console.log("Live.js doesn't support the file protocol. It needs http.");
})();


R/raster.R
#' Print trellis plot to servr
#'
#' @param x trellis object
#' @param \ldots  additional parameters
#' @S3method print trellis
#' @import htmltools
#' @importFrom digest digest
print.trellis <- function(x, ...) {
  print_graphics(x)
}

#' Print ggplot2 plot to servr
#'
#' @param x ggplot object
#' @param \ldots  additional parameters
#' @S3method print ggplot
print.ggplot <- function(x, ...) {
  print_graphics(x)
}

print_graphics <- function(x) {

  graphics_opt <- getOption("rmote_graphics", FALSE)

  if(is_rmote_on() && graphics_opt && no_other_devices()) {
    message("serving graphics through rmote")

    output_dir <- file.path(get_server_dir(), "plots")
    if(!file.exists(output_dir))
      dir.create(output_dir, recursive = TRUE)

    opts <- getOption("rmote_device")
    if(is.null(opts)) {
      rmote_device()
      opts <- getOption("rmote_device")
    }

    if(is.null(opts$filename)) {
      plot_base <- digest::digest(x)
    } else {
      plot_base <- opts$filename
    }
    file <- file.path("plots", paste0(plot_base, ".", opts$type))
    ofile <- file.path(output_dir, paste0(plot_base, ".", opts$type))

    cur_type <- opts$type
    if(opts$type == "png") {
      ww <- opts$width
      hh <- opts$height
      if(opts$retina) {
        opts$width <- opts$width * 2
        opts$height <- opts$height * 2
        opts$res <- 150
      }
      html <- tags$html(
        tags$head(tags$title(paste("raster plot:", plot_base))),
        tags$body(tags$img(src = file,
          width = paste0(ww, "'x"), height = paste0(hh, "'x"))))
      opts$type <- NULL
      opts$retina <- NULL
      opts$filename <- ofile
      if(capabilities("cairo"))
        opts$type <- "cairo-png"
      do.call(png, opts)
    } else if(opts$type == "pdf") {
      html <- tags$html(
        tags$head(tags$title(paste("raster plot:", plot_base))),
        tags$body(tags$a(href = file, "pdf", target = "_blank")))
      opts$type <- NULL
      opts$file <- ofile
      do.call(pdf, opts)
    }

    if(inherits(x, c("trellis", "ggplot"))) {
      if(inherits(x, "trellis"))
        getFromNamespace("print.trellis", "lattice")(x)
      if(inherits(x, "ggplot"))
        getFromNamespace("print.ggplot", "ggplot2")(x)
      dev.off()

      res <- write_html(html)

      # make thumbnail
      if(is_history_on())
        make_raster_thumb(res, cur_type, opts, ofile)

    } else if(inherits(x, "base_graphics")) {
      message("when finished with plot commands, call plot_done()")
      options(rmote_baseplot = list(html = html, ofile = ofile,
        cur_type = cur_type, opts = opts))
    }

    return()
  } else {
    if(inherits(x, "trellis"))
      return(getFromNamespace("print.trellis", "lattice")(x))
    if(inherits(x, "ggplot"))
      return(getFromNamespace("print.ggplot", "ggplot2")(x))
  }
}

#' Set device parameters for traditional grahpics plot output
#'
#' @param type either "png" or "pdf"
#' @param filename optional name for file (should have no extension and no directories)
#' @param retina if TRUE and type is "png", the png file will be plotted at twice its size and shown at the original size, to look better on high resolution displays
#' @param \ldots parameters passed on to either \code{\link[grDevices]{png}} or \code{\link[grDevices]{pdf}} (such as width, height, etc.)
#' @export
rmote_device <- function(type = c("png", "pdf"), filename = NULL, retina = TRUE, ...) {

  type <- match.arg(type)
  opts <- list(...)
  opts$type <- type
  opts$filename <- filename
  if(type == "png") {
    opts$retina <- retina
    if(is.null(opts$width))
      opts$width <- 480
    if(is.null(opts$height))
      opts$height <- 480
  }

  options(rmote_device = opts)
}

make_raster_thumb <- function(res, cur_type, opts, ofile) {
  message("making thumbnail")
  fbase <- file.path(get_server_dir(), "thumbs")
  if(!file.exists(fbase))
    dir.create(fbase)
  nf <- file.path(fbase, gsub("html$", "png", basename(res)))
  if(cur_type == "pdf") {
    opts <- list(filename = nf, width = 300, height = 150)
    if(capabilities("cairo"))
      opts$type <- "cairo-png"
    do.call(png, opts)
    getFromNamespace("print.trellis", "lattice")(text_plot("pdf file"))
    dev.off()
  } else {
    suppressMessages(make_thumb(ofile, nf, width = opts$width, height = opts$height))
  }
}



